ipa-core/src/protocol/ipa_prf/shuffle/sharded.rs

```diff
@@ -0,0 +1,522 @@
+/// that the data it holds is small (a pair of shared values) which is not true for impressions and
+/// conversions that are fed into the shuffle.
+///
+/// [`ShuffleShare`] and [`Shuffleable`] are added to bridge the gap. They can be implemented for
+/// arbitrary structs as long as `Add` operation can be defined on them.
+trait ShuffleShare: Sendable + Clone + FromRandom + Add<Output = Self> {}
+
+impl<V: Sendable + Clone + FromRandom + Add<Output = Self>> ShuffleShare for V {}
+
+/// Trait for shuffle inputs that consists of two values (left and right).
+trait Shuffleable: Send + 'static {
+    type Share: ShuffleShare;
+
+    fn left(&self) -> Self::Share;
+    fn right(&self) -> Self::Share;
+
+    fn new(l: Self::Share, r: Self::Share) -> Self;
+}
+
+impl<V: SharedValue + FromRandomU128> Shuffleable for AdditiveShare<V> {
+    type Share = V;
+
+    fn left(&self) -> Self::Share {
+        ReplicatedSecretSharing::left(self)
+    }
+
+    fn right(&self) -> Self::Share {
+        ReplicatedSecretSharing::right(self)
+    }
+
+    fn new(l: Self::Share, r: Self::Share) -> Self {
+        ReplicatedSecretSharing::new(l, r)
+    }
+}
+
+/// Sharded shuffle as performed by shards on H1.
+async fn h1_shuffle<I, S, C>(ctx: C, shares: I) -> Result<Vec<S>, crate::error::Error>
+where
+    I: IntoIterator<Item = S>,
+    I::IntoIter: Send + ExactSizeIterator,
+    C: ShardedContext,
+    S: Shuffleable,
+{
+    // Generate X_1 = perm_12(left ⊕ right ⊕ z_12).
+    let x1 = ctx
+        .narrow(&ShuffleStep::Permute12)
+        .mask_and_shuffle::<_, S::Share>(
+            Direction::Right,
+            shares.into_iter().map(|share| share.left() + share.right()),
+        )
+        .await?;
+
+    // Generate X_2 = perm_31(x_1 ⊕ z_31) and reshard it using the randomness
+    // shared with the left helper.
+    let x2 = ctx
+        .narrow(&ShuffleStep::Permute31)
+        .mask_and_shuffle(Direction::Left, x1)
+        .await?;
+
+    // X2 is masked now and cannot reveal anything to the helper on the right.
+    ctx.narrow(&ShuffleStep::LeftToRight)
+        .send_all(x2, Direction::Right)
+        .await?;
+
+    // H1 does not know anything about C. In a non-sharded world, the cardinality of C
+    // can be derived from input, sharding makes it impossible to derive because shares
+    // are not distributed evenly across shards. Thus, each shard on H2 must inform H1 peer
+    // about the size of C, so H1 can use PRSS to set its own shares.
+    let sz = ctx
+        .narrow(&ShuffleStep::Cardinality)
+        .recv_word(Direction::Right)
+        .await?;
+
+    // set our shares
+    let ctx = ctx.narrow(&ShuffleStep::PseudoRandomTable);
+    Ok((0..sz)
+        .map(|i| {
+            // This may be confusing as paper specifies Ã and B̃ as independent tables, but
+            // there is really no reason to generate them using unique PRSS keys.
```

Comment:
The prss key consists of two AES keys left and right. We need to make sure that the party who receives the masked shares does not have access to the Prss key and therefore looks uniform to the party.

Reply:
based on my analysis of the protocol, it should be fine because

$\tilde{A}$ is known to $H_1$ and $H_3$
$\tilde{B}$ is known to $H_1$ and $H_2$

* $H_1$ never exposes $\tilde{A}$ and $\tilde{B}$
* $H_2$ uses $\tilde{B}$ to mask data sent to $H_3$ and generates $\tilde{B}$ using PRSS shared with $H_1$. As stated above, $H_3$ does not know $\tilde{B}$
* $H_3$ uses $\tilde{A}$ to mask data sent to $H_2$ and generates $\tilde{A}$ using PRSS shared with $H_1$. $H_2$ does not know  $\tilde{A}$

My understanding that if the original AES key was expanded using the same info context does not make it insecure, because the original keys are not known to the other pair